fix(job-executor): Adding helpful debug logs to improve job acquisition and job execution logging. Issue#2666.

prakashpalanisamy opened the pull request:
This code change adds additional logging to Camunda Job Executor and scoped to the Job Acquisition and Job Execution. 
![je](https://github.com/camunda/camunda-bpm-platform/assets/12576777/74bfbff8-4be3-40ca-afee-2aa618cea27c)

https://github.com/camunda/camunda-bpm-platform/issues/2666

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/camunda/camunda-bpm-platform?pullRequest=4217) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/camunda/camunda-bpm-platform?pullRequest=4217) before we can accept your contribution.<br/><hr/>**Palanisamy, Prakash** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/camunda/camunda-bpm-platform?pullRequest=4217) it.</sub>

yanavasileva commented:
Hi @prakashpalanisamy,

Thank you for raising this.
We will have a look at it and get back to you.

Best,
Yana

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/SequentialJobAcquisitionRunnable.java:
```diff
@@ -157,6 +157,8 @@ protected AcquiredJobs acquireJobs(
 
     int numJobsToAcquire = acquisitionStrategy.getNumJobsToAcquire(currentProcessEngine.getName());
 
+    LOG.jobsToAcquire(currentProcessEngine.getName(),numJobsToAcquire);
```
Formatting: 
```suggestion
    LOG.jobsToAcquire(currentProcessEngine.getName(), numJobsToAcquire);
```

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/SequentialJobAcquisitionRunnable.java:
```diff
@@ -173,6 +175,7 @@ protected AcquiredJobs acquireJobs(
     jobExecutor.logAcquisitionFailureJobs(currentProcessEngine, acquiredJobs.getNumberOfJobsFailedToLock());
 
     LOG.acquiredJobs(currentProcessEngine.getName(), acquiredJobs);
+    LOG.failedAcquisitions(currentProcessEngine.getName(),acquiredJobs);
```
Formatting: 
```suggestion
    LOG.failedAcquisitions(currentProcessEngine.getName(), acquiredJobs);
```

yanavasileva reviewed (commented):
@prakashpalanisamy, nice work!
I left a few comments for the code formatting (check all of the code for missing spaces) and covering the non-Spring use cases.

In addition to the changes, could you have have a look at adding a few test cases for the logging. The logging rule can be helpful for them. You can see examples here:
* https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/test/java/org/camunda/bpm/engine/test/jobexecutor/JobExecutorAcquireJobsForPriorityRangeTest.java#L164-L176
* https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/test/java/org/camunda/bpm/engine/test/jobexecutor/JobExceptionLoggingTest.java#L77-L96

yanavasileva commented:
Closing due to inactivity. The PR can be reopened again when you get back to the topic.

prakashpalanisamy commented:
Hi @yanavasileva Requesting you to please reopen the PR. I would like to continue the contribution to this PR. Thank you!

prakashpalanisamy commented:
@yanavasileva , hope you are doing good. I have updated the PR with additional commits. Can you please take a look into it when you get some time. Thanks!!

yanavasileva commented:
I will try to have a look and provide input later this week.

prakashpalanisamy commented:
@yanavasileva Please let me know on the comments once you have finished reviewing, I will have them addressed. Thank you! :)

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/JobExecutorLogger.java:
```diff
@@ -217,4 +217,52 @@ public ProcessEngineException jobExecutorPriorityRangeException(String reason) {
+            "Jobs currently in queue to be executed for the process engine '{}' is {} out of the max queue size : {}", processEngine, numJobsInQueue, maxQueueSize);
+  }
+
+  public void availableThreadsCalculationError() {
+    logWarn(
+            "039",
+            "Arithmetic exception occurred while computing remaining available thread count for logging.");
+  }
+
+  public void totalQueueCapacityCalculationError() {
+    logWarn(
+            "040",
+            "Arithmetic exception occurred while computing total queue capacity for logging.");
+  }
```
❌ Change severity to `debug` since the related logs are also logged with it + formatting:
```suggestion
  public void availableThreadsCalculationError() {
    logDebug("039", "Arithmetic exception occurred while computing remaining available thread count for logging.");
  }

  public void totalQueueCapacityCalculationError() {
    logDebug("040", "Arithmetic exception occurred while computing total queue capacity for logging.");
  }
```

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/JobExecutorLogger.java:
```diff
@@ -217,4 +217,52 @@ public ProcessEngineException jobExecutorPriorityRangeException(String reason) {
+  public void numJobsInQueue(String processEngine, int numJobsInQueue, int maxQueueSize) {
+    logDebug(
+            "038",
+            "Jobs currently in queue to be executed for the process engine '{}' is {} out of the max queue size : {}", processEngine, numJobsInQueue, maxQueueSize);
```
🔧 Typo:
```suggestion
            "Jobs currently in queue to be executed for the process engine '{}' are {} out of the max queue size : {}", processEngine, numJobsInQueue, maxQueueSize);
```

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/JobExecutor.java:
```diff
@@ -171,13 +171,30 @@ public void logAcquisitionFailureJobs(ProcessEngineImpl engine, int numJobs) {
     }
   }
 
+  public void logJobExecutionInfo(ProcessEngineImpl engine, int executionQueueSize, int executionQueueCapacity, int maxExecutionThreads, int activeExecutionThreads) {
```
❓ Could you also have a look at https://github.com/camunda/camunda-bpm-platform/blob/master/engine-cdi/core/src/main/java/org/camunda/bpm/engine/cdi/impl/ManagedJobExecutor.java?

yanavasileva reviewed (changes requested):
❌ Formatting is off almost everywhere. You can find the latest formatters here:
https://github.com/camunda/camunda-bpm-platform/tree/master/settings
👍 Changes look good so far, thank you for considering my feedback.
I added two more notes below. I will validate the added tests, run some further scenarios, and get back to you with my findings.

Thank you for your patience.

yanavasileva reviewed (commented):
Adding some further adjustments related to the tests.

yanavasileva commented on engine/src/test/java/org/camunda/bpm/engine/test/jobexecutor/JobExecutionLoggingTest.java:
```diff
@@ -0,0 +1,190 @@
+    List<ILoggingEvent> filteredLogList = loggingRule.getFilteredLog("Jobs currently in queue to be executed for the process engine 'default' is 2 out of the max queue size : 2");
+
+    // then 3 instances of filled queue logs will be available as there is 2 additional threads possible to reach max-pool-size
+    assertThat(filteredLogList.size()).isEqualTo(3);
```
This check if flaky.

yanavasileva commented on engine/src/test/java/org/camunda/bpm/engine/test/jobexecutor/JobExecutionLoggingTest.java:
```diff
@@ -0,0 +1,190 @@
+    List<ILoggingEvent> filteredLogList = loggingRule.getFilteredLog("Jobs currently in execution for the process engine 'default' : 3");
+
+    // then one count of 'three jobs in execution' will be available as 2 out of 5 jobs should be queued.
+    assertThat(filteredLogList.size()).isEqualTo(1);
```
This check is flaky.

yanavasileva reviewed (changes requested):
❌ I have ran the tests a couple of times but they are unstable. Could you please iterate on them to make them reliable?

yanavasileva commented:
@prakashpalanisamy, heads up, we approach the code freeze of the minor release.
❗ If you want the feature to be included in 7.22.0, please apply the feedback by **16th of September**.

prakashpalanisamy commented:
> @prakashpalanisamy, heads up, we approach the code freeze of the minor release. ❗ If you want the feature to be included in 7.22.0, please apply the feedback by **16th of September**.

Sure @yanavasileva I'll take this up on priority. Was tied up due to personal matters. I may need your inputs on the Tests though.  I observed that some of the new logs and already present logs were getting into the log file while the JE gets loaded. sometime twice. So the number of occurrences were varying on my system as well. Hence I was opting for greater than checks. Will test that once more and reach out to you with specifics for directions. Thank you for assisting me with this PR :)

prakashpalanisamy commented:
@yanavasileva - quick question on the formatting. When I apply the formatting, it also suggest changes to prior code that other authors have contributed in the past, on the files that I have changes in. Do you suggest that I apply them along with my changes? else only to the lines that Im contributing and ignoring other formatting suggestions? Thanks!!

yanavasileva commented:
@prakashpalanisamy,

You can mark only your code and format only that. This will be easier to review by me.

prakashpalanisamy commented:
@yanavasileva - I need your directions for the ManagedJobExecutor changes. Since it will be using the app servers' thread pool, should I set it up in wildfly distribution with ManagedExecutorService configuration? if so, any pointers on how to do the configurations? If you could point to any reference links, it will help me. 
Basically, once i put the code in, I want to get it up and running and test manually with some process of mine having async scenarios with java delegates.

yanavasileva commented:
I won't be able to provide input on this, we can create a follow up ticket to explore further `ManagedJobExecutor`. Feel free to apply the rest of the feedback.

For the tests, I also saw that the log entries can vary. Maybe you can do the assertion for minimum occurrences?

prakashpalanisamy commented:
@yanavasileva - Hope you are doing good. Please find the latest commits having addressed the review comments. On the test cases front, I have made some changes. Please let me know if those looks okay for you or else if changes are needed. Thank you for the patience. Apologies for the unavoidable delays in the push.

yanavasileva commented:
The changes looks good. Thank you for your contribution, shortly I will be merging the changes to master.
I had to apply a fix in the test so I am closing this PR in favour of https://github.com/camunda/camunda-bpm-platform/pull/4558

prakashpalanisamy commented:
Thank you @yanavasileva for the support!!
